ffi/src/winscard/system_scard/context.rs

```diff
@@ -0,0 +1,679 @@
+use std::ptr::{null, null_mut};
+#[cfg(target_os = "windows")]
+use std::slice::from_raw_parts;
+
+#[cfg(target_os = "windows")]
+use ffi_types::winscard::functions::SCardApiFunctionTable;
+#[cfg(target_os = "windows")]
+use ffi_types::winscard::{ScardContext, ScardHandle};
+use uuid::Uuid;
+use winscard::winscard::{
+    DeviceTypeId, Icon, Protocol, ProviderId, ReaderState, ScardConnectData, ScardScope, ShareMode, WinScardContext,
+};
+use winscard::{Error, ErrorKind, WinScardResult};
+
+use super::{parse_multi_string_owned, SystemScard};
+#[cfg(not(target_os = "windows"))]
+use crate::winscard::pcsc_lite::functions::PcscLiteApiFunctionTable;
+#[cfg(not(target_os = "windows"))]
+use crate::winscard::pcsc_lite::{initialize_pcsc_lite_api, ScardContext, ScardHandle};
+
+pub struct SystemScardContext {
+    h_context: ScardContext,
+    #[cfg(target_os = "windows")]
+    api: SCardApiFunctionTable,
+    #[cfg(not(target_os = "windows"))]
+    api: PcscLiteApiFunctionTable,
+}
+
+impl SystemScardContext {
+    #[allow(dead_code)]
+    pub fn establish(scope: ScardScope) -> WinScardResult<Self> {
+        let mut h_context = 0;
+
+        #[cfg(target_os = "windows")]
+        let api = super::init_scard_api_table()?;
+        #[cfg(not(target_os = "windows"))]
+        let api = initialize_pcsc_lite_api()?;
+
+        // SAFETY: This function is safe to call because the `scope` parameter value is type checked
+        // and `*mut h_context` can't be `null`.
+        try_execute!(
+            unsafe { (api.SCardEstablishContext)(scope.into(), null_mut(), null_mut(), &mut h_context) },
+            "SCardEstablishContext failed"
+        )?;
+
+        if h_context == 0 {
+            return Err(Error::new(
+                ErrorKind::InternalError,
+                "Can not establish context: SCardEstablishContext did not set the context handle",
+            ));
+        }
+
+        Ok(Self { h_context, api })
+    }
+}
+
+impl Drop for SystemScardContext {
+    fn drop(&mut self) {
+        // SAFETY: This function is safe to call because the `self.h_context` is always a valid handle.
+        if let Err(err) = try_execute!(
+            unsafe { (self.api.SCardReleaseContext)(self.h_context) },
+            "SCardReleaseContext failed"
+        ) {
+            error!(?err, "Can not release the scard context");
+        }
+    }
+}
+
+impl WinScardContext for SystemScardContext {
+    fn connect(
+        &self,
+        reader_name: &str,
+        share_mode: ShareMode,
+        protocol: Option<Protocol>,
+    ) -> WinScardResult<ScardConnectData> {
+        // SAFETY:
+        // https://doc.rust-lang.org/std/ffi/struct.CString.html#method.new
+        // > This function will return an error if the supplied bytes contain an internal 0 byte.
+        //
+        // The Rust string slice cannot contain 0 bytes. So, it's safe to unwrap it.
+        let c_string = CString::new(reader_name).expect("Rust string slice should not contain 0 bytes");
```

Comment:
style: `SAFETY:` comments should be used for `unsafe` blocks only. Of coure, comments explaining why the code does not panic are very welcomed too, but causing a panic is not unsafe.